src/psyclone/tests/dynamo0p3_test.py

```diff
@@ -6531,3 +6601,459 @@ def test_stub_generate_with_anyw2():
+    psy = PSyFactory("dynamo0.3").create(info)
+    schedule = psy.invokes.invoke_list[0].schedule
+    stencil_halo_exchange = schedule.children[1]
+    assert stencil_halo_exchange._compute_stencil_type() == "region"
+
+
+def test_halo_compute_halo_internal_error(monkeypatch):
+    '''Check that we raise an exception if the compute_halo_read_info method in
+    dynhaloexchange does not find any read dependencies. This should
+    never be the case. We use monkeypatch to force the exception to be
+    raised'''
+    _, invoke_info = parse(os.path.join(BASE_PATH, "1_single_invoke.f90"),
+                           api="dynamo0.3")
+    psy = PSyFactory("dynamo0.3").create(invoke_info)
+    schedule = psy.invokes.invoke_list[0].schedule
+    halo_exchange = schedule.children[0]
+    field = halo_exchange.field
+    monkeypatch.setattr(field, "forward_read_dependencies", lambda fs=None: [])
+    with pytest.raises(GenerationError) as excinfo:
+        halo_exchange._compute_halo_read_info()
+    assert ("Internal logic error. There should be at least one read "
+            "dependence for a halo exchange") in str(excinfo.value)
+
+
+def test_halo_exchange_one_backward_dependence(monkeypatch):
+    '''Check that an internal error is raised if a halo exchange returns
+    with more than one write dependency. It should only ever be 0 or 1.'''
+    _, invoke_info = parse(os.path.join(BASE_PATH, "1_single_invoke.f90"),
+                           api="dynamo0.3")
+    psy = PSyFactory("dynamo0.3").create(invoke_info)
+    schedule = psy.invokes.invoke_list[0].schedule
+    halo_exchange = schedule.children[0]
+    field = halo_exchange.field
+    #
+    monkeypatch.setattr(field, "backward_write_dependencies",
+                        lambda fs=None: [1, 1])
+    with pytest.raises(GenerationError) as excinfo:
+        halo_exchange._compute_halo_write_info()
+    assert ("Internal logic error. There should be at most one "
+            "write dependence for a halo exchange. Found "
+            "'2'") in str(excinfo.value)
+    #
+    monkeypatch.setattr(field, "backward_write_dependencies",
+                        lambda fs=None: [])
+    assert halo_exchange._compute_halo_write_info() == []
+
+
+def test_halo_exchange_backward_dependence_no_call(monkeypatch):
+    '''Check that an internal error is raised if a halo exchange
+    write dependency is not a call.'''
+    _, invoke_info = parse(os.path.join(BASE_PATH,
+                                        "14.9_halo_different_stencils.f90"),
+                           api="dynamo0.3")
+    psy = PSyFactory("dynamo0.3").create(invoke_info)
+    schedule = psy.invokes.invoke_list[0].schedule
+    halo_exchange = schedule.children[1]
+    field = halo_exchange.field
+    write_dependencies = field.backward_write_dependencies()
+    write_dependency = write_dependencies[0]
+    monkeypatch.setattr(write_dependency, "call",
+                        lambda fs=None: halo_exchange)
```

Comment:
Should that be `self=None` rather than `fs=None` as the argument to the lambda, or maybe even an un-named argument? Also, why does the error message below contain "<type 'function'>" instead of something related to `HaloExchange`?